zim/gui/widgets.py

```diff
@@ -2060,35 +2060,44 @@ def _fill_completion_for_anchor(self, path, prefix, text):
 			pass
 
 	def _fill_completion_any(self, path, text):
+
+		# Don't match single-character completions
+		if len(text) < 2:
+			return;
+
 		#print "COMPLETE ANY", path, text
 		# Complete all matches of "text"
 		# start with children, than peers, than rest of tree
 		# if path == ":" don't use child notation
 		completion = self.get_completion()
 		completion.set_match_func(gtk_entry_completion_match_func, 1)
 
-		# TODO: use SQL to list all at once instead of walking and filter on "text"
-		#       do better sorting as well ?
+		# TODO: do better sorting?
 
 		def relative_link(target):
 			href = self.notebook.pages.create_link(path, target)
 			return href.to_wiki_link()
 
 		model = completion.get_model()
-		lowertext = text.lower()
-		childpos, peerpos = 0, 0
-		for p in self.notebook.pages.walk():
-			if lowertext in p.basename.lower():
-				link = relative_link(p)
-				if link.startswith('+'):
-					model.insert(childpos, (link, p.basename))
-					childpos += 1
-					peerpos += 1
-				elif not ':' in link:
-					model.insert(peerpos, (link, p.basename))
-					peerpos += 1
-				else:
-					model.append((link, p.basename))
+		count, childpos, peerpos = 0, 0, 0
+		for p in self.notebook.pages.search_pagename_substring(text):
+
+			count=count+1;
+			# BENCHMARK: 1 second of UI hang for every 100 entries.
+			if count > 300:
+				logger.debug("truncating long results list to prevent GTK hang")
```

Comment:
Would remove these kind of debug messages before merge - they are useful while developing, but later on only add noise to the error log